### DISCUSSION TITLE
Is there a good way to solve naming conflicts with rxjs?

### DISCUSSION THREAD

**Discussion by hillin** [iter-tools/iter-tools#472] [General]
I fully understand this is not a problem which should be addressed by neither iter-tools nor rxjs, but it's a annoyance that truly exist: both libraries share too many operator names, especially `map` and `filter` which are my bread and butter of daily work from both libs.

I also understand I can always give them aliases as I import:
```typescript
import { map as mapRx } from 'rxjs';
import { map as mapIt } from 'iter-tools-es';
```
or simply import as namespaces:
```typescript
import * as rx from 'rxjs';
import * as it from 'iter-tools-es';
```
of which both techniques I use to avoid the naming conflicts; but from time to time, I still find myself wasting time checking why my newly written observable has wavy lines under its 100-lines long body, only to realize five minutes later that I'm using iter-tools' `map` instead of the rxjs one, and that's because when I simply type `m`,  the smartass vscode IDE automatically imports it for me without myself ever noticed.

I wonder is there an elegant way to solve this problem, once and for all preferably?

**Comment by conartist6** [comment]
At one point in time I had the idea of creating a [babel macro](https://github.com/kentcdodds/babel-plugin-macros) which could solve the problem of it being a pain in the butt to keep writing `as` clauses for your imports:
```js
import { Sequence } from 'iter-tools/sequence.macro';
import { map } from 'rx';

[...new Sequence(iterable).map(() => /* ... */).filter(() => /* ... */)];
```

During the build step the macro would transform that code to this:
```js
import { execPipe, map as mapIter, filter } from 'iter-tools';
import { map } from 'rx';

[...execPipe(iterable, mapIter(() => /* ... */), filter(() => /* ... */))];
```
...and of course the macro could use babel's scope analysis to resolve the namespace conflicts for you then.

Want to build it?

**Reply by conartist6** [reply]
I have a bunch of the parts done so it shouldn't be too hard.

- There's already `explode.macro` as part of iter-tools, and contains similar logic for replacing a macro import with imports from the library itself.
- The methods in the library already contain the relevant metadata: methods which are transforms or reductions (valid operations on a sequence) are tagged, and there are existing patterns for building up data objects containing metadata about the library, such as to generate `API.md`
- Code generation is already part of the project, and I'm actually spending most of my time working on tools for code generation and codemodding.

**Reply by conartist6** [reply]
Also typescript doesn't understand the concept of macros really. You just write normal typedefs for the module. In this case you'd have to use more AST hackery to get it to take all the normal typedefs and make it look like they belong to the non-existent Sequence class.

**Reply by conartist6** [reply]
And of course I should mention that for anyone writing plain node code there's no build step, and thus this won't work. I'm working on a solution to that problem as well with the [macrome](https://github.com/macrome-js/macrome) build system, but right now I have a lot of stuff up in the air and it hasn't quite started to come down yet.

**Reply by hillin** [reply]
Sounds interesting! I'm not quite familiar with babel (I mostly work with things provided by angular), what should I do?

**Comment by conartist6** [comment]
From what you've said there can be no possible change to iter-tools that would solve wrong-import part of your problem, since your problem is the VSCode thinks you're trying to use iter-tools when you're not.

To fully remove the ambiguity you'd have to make wrappers around iter-tools and rxjs which did the name prefixing so that you'd always type out `rxMap` or `itMap` but never just `map`. It seems to me like the correct solution to that problem is to fix VSCode, especially since that fix will benefit all developers (including me!).

**Reply by hillin** [reply]
> It seems to me like the correct solution to that problem is to fix VSCode

It doesn't seem to be a VSCode issue either, because it actually does show a popup and allows me to choose between the different functions, it's just when you type very fast with your muscle memory, it's too easy to be ignored.

Honestly I didn't hope to find an answer to this problem, it's something that no one was wrong but simply exists in a frustrating shape. I tried to make this post several times and gave up just because every time I quickly calm down and reasoned to my self this would never be constructive, it most likely would end up just a complaining. And yet there is another occurrence and I couldn't help ;-)

**Reply by conartist6** [reply]
If something keeps tripping you up, it's a problem worth looking at fixing!! As I said, it's tripped me up quite a few times as well. And as you say, every time it's easier to go back and remove the things I didn't mean to add rather than try to fix the problem. If everyone takes this approach, it will never change! Let's take this discussion to the VSCode forums and see if anyone there knows of configurations or plugins or PRs or other suggestions for how to make the behavior less annoying.

**Reply by hillin** [reply]
I admire your passion... but I failed to figure out a way to express this problem to the VSCode guys. They even have a setting that allows you to disable auto-importing, but I wholeheartedly don't want to turn it off...

**Reply by conartist6** [reply]
Good! The first step to fixing any problem is understand what it is.

I think the problem is just some extremely poor UX. If I press the keys `m`, `a`, `p`, `(` an auto-import is executed. That's just bonkers! Often times I was just typing and didn't even notice the suggestion list.

**Reply by conartist6** [reply]
Ah look! You can change that! The setting for it is called: "Editor: Accept Suggestion On Commit Character". After turning it off you can now type `map(` and nothing happens -- you just have a normal unresolved import. You can then click on the unresolved symbol, click on the light bulb above it, and select the correct source for the import from the dropdown.

**Reply by hillin** [reply]
> I think the problem is just some extremely poor UX

Well I'd say it's a feature, it works great for me in other scenarios, and if I'm only working with rxjs or iter-tools, it works very well too, so I'd not complain about it. I'm constantly dealing with files with tens of imports, so fairly speaking it has more pros than cons to me, which as I said, I don't want to turn it off. That being said, this can be a solution to my problem.

**Comment by conartist6** [comment]
OK, new answer based on [this comment](https://github.com/iter-tools/iter-tools/discussions/472#discussioncomment-2871589) to promote visibility:

One way that you can improve the behavior is to disable VSCode's "Accept Suggestion On Commit Character" setting. This is what makes the editor do an auto import the moment you type the letters `map(`. `(` is a commit character for javascript, and so without even meaning to interact with the suggestion box (or maybe even without realizing you did) you have added an import to your file. If you don't allow commit characters to accept suggestions then you'll just end up with `map()` and the `map` identifier will be unresolved, which is perfect because then you can use the "Quick fix" functionality to resolve it. Tip: look quick fix up in the command palette to see what the shortcut is for it! I'm on OSX so it's `cmd .`.
